### ISSUE TITLE
Publishing to Maven Central

### ISSUE THREAD

**Issue by lemire** [RoaringBitmap/RoaringBitmap#749]
We have stopped publishing to Maven Central for the time being as our existing configuration stopped working.

I am calling for help. If someone would contribute a CI workflow as well as a gradle configuration that allows us to publish our releases automatically, that would be great.

Please see also https://github.com/RoaringBitmap/RoaringBitmap/issues/748

**Comment by ChicChip** [comment]
how can i import bsi jar from jitpack

**Comment by lhotari** [comment]
@lemire Do you need more help in addressing this issue? I noticed that #748 was closed. Is there anything else that prevents publishing to Maven Central? I can see in https://repo1.maven.org/maven2/org/roaringbitmap/RoaringBitmap/ that the latest release available in Maven Central is 1.3.0, however the latest release of RoaringBitmap is 1.3.16.

**Comment by lemire** [comment]
@lhotari We have automated releases (JitPack and GitHub packages). We are inviting pull requests for other platforms, including Maven Central. If you'd like to fix this issue, a pull request is invited. The process should be automated, and not require constant maintenance.

For the time being, if you want up-to-date releases, we encourage you to adopt either JitPack or GitHub packages. We thoroughly document the usage (see README.md).

**Comment by lhotari** [comment]
> @lhotari We have automated releases (JitPack and GitHub packages). We are inviting pull requests for other platforms, including Maven Central. If you'd like to fix this issue, a pull request is invited. The process should be automated, and not require constant maintenance.
> 
> For the time being, if you want up-to-date releases, we encourage you to adopt either JitPack or GitHub packages. We thoroughly document the usage (see README.md).

@lemire I could work on making the Maven Central publishing work again. A lot of Java projects use Maven Central and that's the most convenient way also for Apache Pulsar. 
I noticed a comment "As of Oct 22 2024 this fails." in this commit 3ec50b66 . Do you happen to have a problem description? I can also make progress if you don't have the problem description available.

**Comment by lemire** [comment]
> I also noticed that Gradle is an old version, 6.9.4 (released in Feb 2023) . I could first help get the build migrated to use a recent Gradle version.

Pull request invited.

Note however that the build configuration should not depend on Sonatype proprietary bits. Otherwise, when Sonatype changes its processes, it may break out builds which is not acceptable. Sonatype  components should be separate so that we do not suffer from vendor lock-in.

E.g., we should use on-release automated triggers:

https://github.com/RoaringBitmap/RoaringBitmap/blob/master/.github/workflows/releases.yml

> Do you happen to have a problem description? I can also make progress if you don't have the problem description available.

Sonatype rejected the release which broke our release process. Further, it never worked properly. The release and publish process needs to be automated, like JitPack or GitHub packages.

**Comment by blacelle** [comment]
Some additional feedback:
- I previously worked on RoaringBitmap m2central deployment, though it felt terrible to me as the related code is now in Gradle+Kotlin (and I'm much more confortable with maven, Gradle is OK-ish, Gradle+Kotlin felt too much as RoaringBitmap is my only use of it).
- I manage other libraries deployements to m2-central (with mvn config) and it works just fine (i.e. I do not encounter the large set of issues managed through RoaringBitmap). I wonder if our difficulties are are due to pure-luck, day-time related (e.g. I tend to release in hours when Sonatype servers are fine), maven vs gradle related, or anything else.

A bit of history:
- https://github.com/RoaringBitmap/RoaringBitmap/issues/197 Many issues around `pom.xml` and `settings.xml` to deploy to Sonatype. Got fixed.
- https://github.com/RoaringBitmap/RoaringBitmap/issues/201 Weird issues about some jars missing from Sonatype (mvn)
- https://github.com/RoaringBitmap/RoaringBitmap/issues/345 Random Sonatype issues, and referring to migrating to Gradle.
- https://github.com/RoaringBitmap/RoaringBitmap/issues/360 Issues syncing Bintray with Sonatype
- https://github.com/RoaringBitmap/RoaringBitmap/issues/454 Going back to direct deployement to Sonatype.
- https://github.com/RoaringBitmap/RoaringBitmap/issues/478 Change of Sonatype server.
- https://github.com/RoaringBitmap/RoaringBitmap/commit/3ec50b668e2fa8a9bee37d0631712260df711f48 Drops autoamtic deployement to m2central.

**Comment by blacelle** [comment]
> Note however that the build configuration should not depend on Sonatype proprietary bits. Otherwise, when Sonatype changes its processes, it may break out builds which is not acceptable. Sonatype components should be separate so that we do not suffer from vendor lock-in.

I understand your point, through I disagree with it. I suppose it is based on RoaringBitmap (poor) experience of Sonataype, which is not the general experience.

> we should use on-release automated triggers:

Yes. And if it were automated, it would be a bit less of an issue to depends on Sonatype bits. Worst case, the retry process could be automated, or reduced to a single click.

> I could first help get the build migrated to use a recent Gradle version. It could be easier to follow recent Gradle best practices when a recent version of Gradle is used for the build. I used to work in the Gradle build tool team in 2015-2016 so I think I could help with this task.

This would be very helpful @lhotari. I can provide help around the matter (e.g. on the publication part, post Gradle upgrade; and/or the CI part).

(I personally believe very important to keep publishing to m2central).

**Comment by lhotari** [comment]
Thanks for the additional context @lemire and @blacelle. 

I'll first address the Gradle build upgrade to Gradle 8.x as a preparation, in a separate PR.

For automated publishing from the Gradle build, I'll evaluate the use of [JReleaser](https://jreleaser.org/) which also comes with a Gradle plugin.
JReleaser supports automated releases to multiple destinations. It will be possible to use JReleaser in the Gradle build for both GitHub packages publishing and Maven Central publishing.

**Comment by blacelle** [comment]
Hi @lhotari,

Have you by any change been able to progress on this matter? Thanks

**Comment by lhotari** [comment]
> Hi [@lhotari](https://github.com/lhotari),
> 
> Have you by any change been able to progress on this matter? Thanks

@blacelle  I'm sorry I haven't followed up. I'll add to my list of things to make progress.

**Comment by anandheritage** [comment]
Not quite versed with CI/CD's but let me give it a shot too.

**Comment by anandheritage** [comment]
https://github.com/RoaringBitmap/RoaringBitmap/pull/798 
First upgrading the gradle version. After this I will try to get the pipeline working.
@lemire Help me review this ^^

**Comment by anandheritage** [comment]
Also I have published a snapshot from CI/CD from my forked repo.
Will refine and raise the next PR once the above one is merged.

https://central.sonatype.com/repository/maven-snapshots/io/github/anandheritage/roaringbitmap/1.5.3-SNAPSHOT/roaringbitmap-1.5.3-20250821.100348-1.jar

**Comment by kevinjqliu** [comment]
hey folks, i took a few cycles to look at publishing to maven. 
I was able to publish to my own maven namespace with #820 from [my own fork](https://github.com/kevinjqliu/RoaringBitmap/actions/runs/22163771302/job/64086427998)

See
https://central.sonatype.com/repository/maven-snapshots/io/github/kevinjqliu/roaringbitmap/1.6.1-SNAPSHOT/maven-metadata.xml

This artifact pass CI when tested against Iceberg's CI https://github.com/kevinjqliu/iceberg/pull/11

Could someone please help review #820?

**Comment by lemire** [comment]
PR https://github.com/RoaringBitmap/RoaringBitmap/pull/818 was a total failure.

**Comment by lhotari** [comment]
Regarding https://github.com/RoaringBitmap/RoaringBitmap/issues/749#issuecomment-3928770838

> What is the conceptual flaw?

The problem is that `./gradlew release -Prelease.releaseVersionBump=${{ github.event.inputs.bumpType }}` creates a tag and then sets the version to a snapshot version. The next step `./gradlew publishAggregationToCentralPortal --no-scan` continues in the master branch with the snapshot version commit. One way to solve this would be to have a command to simply reset to the previous commit (which is the revision where the release was tagged) or checkout the release tag. Running `git reset --hard HEAD~` would be one way.

Adding this after "Create release"
```
      - name: Reset to previous commit
        run: git reset --hard HEAD~
```

**Comment by lhotari** [comment]
It's odd that there's https://github.com/RoaringBitmap/RoaringBitmap/commit/0153101f in the commit history, but the other version bumps didn't create a commit to bump the version to a snapshot version after creating the tag.

**Comment by kevinjqliu** [comment]
Iceberg CI is green with `org.roaringbitmap:RoaringBitmap` and version `1.6.9` 😄 
https://github.com/kevinjqliu/iceberg/pull/13

**Comment by steveloughran** [comment]
catching up on this. Are there really two versions which differ only by name?

that's potentially trouble on any FS which doesn't do case independence of its m2 cache (windows).

let's wait and see.

**Comment by lemire** [comment]
@steveloughran Yes. It was unavoidable. It is not going to change anymore.
